### PR TITLE
misc pkgs: various cross fixes in preparation for darwin->linux for 18.09

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

Backport of #46857.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/stdenv/linux/default.nix [hunk 1]
```diff
@@ -324,7 +326,8 @@ in
 
       extraNativeBuildInputs = [ prevStage.patchelf prevStage.paxctl ] ++
         # Many tarballs come with obsolete config.sub/config.guess that don't recognize aarch64.
-        lib.optional (!localSystem.isx86) prevStage.updateAutotoolsGnuConfigScriptsHook;
+        lib.optional (!localSystem.isx86 || localSystem.libc == "musl")
+        prevStage.updateAutotoolsGnuConfigScriptsHook;
```
Comment (matthewbauer) [inline pkgs/stdenv/linux/default.nix]: I can't remember if I asked before but I was wondering if we could just do this on x86 as well? Just for consistency.

Comment (Ericson2314) [reply pkgs/stdenv/linux/default.nix]: @matthewbauer I would always do it on all platforms, but that's a mass rebuild on hydra-cached platforms. I'm taking advantage of MUSL's tier-n+1 status to do this right on master :D.

**Comment by matthewbauer** [comment]
IMO putting this in 18.09 is high risk little reward. Better to just shoot for 19.03 being the version that supports macOS -> Linux compilation. We have enough time to work out the kinks.

**Comment by Ericson2314** [comment]
What's the risk? Not too much is being changed.

**Comment by Ericson2314** [comment]
Oh you mean the overall thing?

**Comment by matthewbauer** [comment]
Yeah I guess if it's not a mass rebuild it's fine but otherwise it seems like there are always random issues with these kind of stdenv changes.

**Pull request merged**